fix: gnome crash when opening a new window while a window on the scratch layer is selected

jtaala opened the pull request:
FIXES #448 
Looks like it also FIXES #427 

Fix for potential mutter change/regression (currently seeing in GNOME 43.1+) when opening a new application while a window in the scratch layer is selected causes a gnome to crash.

See https://github.com/pop-os/shell/issues/1550 for similar issue currently seen in pop-shell extension.

jtaala committed:
Now using "transitions-completed" signal before activating new window that's inserted into the scratch layer.
Fix for potential mutter change/regression (currently seeing in GNOME 43.1. See https://github.com/paperwm/PaperWM/issues/448).

jtaala commented:
Note, if this is caused by a recent regression/change in mutter (which is what I believe atm) - then this change might not be needed if/when mutter fixes whatever is causing this.

hallettj commented:
I can confirm that this fixes the described crash for me. For quite a while I have been able to consistently crash my Gnome session by putting Slack into the scratch layer, and invoking either the "Upload from your computer" or "Create a post" features. (Both of those attempt to open a new window.) The fix is much appreciated!

scottames commented:
> I can confirm that this fixes the described crash for me. For quite a while I have been able to consistently crash my Gnome session by putting Slack into the scratch layer, and invoking either the "Upload from your computer" or "Create a post" features. (Both of those attempt to open a new window.) The fix is much appreciated!

Same here. _Very_ much appreciated!

paradajz commented:
Tried it out on my PC. It fixes this issue indeed.

jtaala commented:
Morning all! @Gelbana @tadfisher @ccope - was speaking with @smichel17 and he mentioned to ping you for code reviews.  Hope eveyone has had a good break (or at least a break from work etc.).

If needed (noting that @hallettj approved this PR previously) when someone has time, could I get you to review this PR?  Fixes an easily reproducible bug that's in Gnome 43 (and likely earlier 40+ versions).

smichel17 commented on tiling.js:
```diff
@@ -2536,7 +2536,15 @@ function insertWindow(metaWindow, {existing}) {
             Scratch.makeScratch(metaWindow);
             if (scratchIsFocused) {
-                Main.activateWindow(metaWindow);
+                /**
+                 * fix for potential mutter change/regression (currently seeing in
+                 * GNOME 43.1.  See https://github.com/paperwm/PaperWM/issues/448).
+                 * Now using "transitions-completed" signal before
+                 * activating new window on scratch layer.
+                 */
+                signals.connectOneShot(actor,'transitions-completed', () => {
+                    Main.activateWindow(metaWindow);
+                });
```
Is this the kind of bug that might pop up again in other places? Does it make sense to pull this out into a helper function, or is it not really re-usable? My main concern is how the accumulation of workarounds like this might make this function hard to understand or change. In other words, where is the right place for this complexity to live?

I have no knowledge of gnome-shell internals and very little of PaperWM, so these are actual questions, not opinions-disguised-as-questions. If this is the right place, that's good; I only want to make sure it was considered.

jtaala replied:
This is a really good question - as far as I can tell, the error/issue is only observed with floating windows and not those managed while tiled in PaperWM.  Given this, and that it's pretty isolated to such a use-case , a helper function would also likely be pretty isolated.  

Certainly I agree though (and it woudn't hurt) - I could implement a helper function for activating windows (e.g. `Tiling.activateWindowAfterTransitions(metaWindow)` or something).  At least I could then add some function comments referencing the issue in one spot etc.).

jtaala replied:
Thanks @smichel17 this one should be good to go now.  Created the function that can be reused to activate window only after transitions are complete.

jtaala committed:
Created function to ease resuse of activating window only after
transitions are complete.  Referenced original issue in funtion
comments.

jtaala committed:
Updated function description to be more correct (re transitions).

smichel17 reviewed (approved):
I'd love if someone else would review this for the gnome-specific bits (confirming that this is the correct workaround), but otherwise the code looks good to me.

jtaala committed:
FIX: now passing actor to activateWindowsAfterTransitions function.

jtaala committed:
Merge remote-tracking branch 'upstream/develop' into fix-issue448-crash-on-new-window-while-scratchwindow-selected

jtaala committed:
Moved to using 'show' signal (which is emitted after actor is
visible/rendered on stage).
It seems 'transitions-completed' signal does work, but was emitted
before the window was rendered, and hence it's workspace property wasn't
assigned and caused an 'workspace' is null error in the logs.

jtaala commented:
> I'd love if someone else would review this for the gnome-specific bits (confirming that this is the correct workaround), but otherwise the code looks good to me.

Thanks @smichel17, I've been digging into this issue (on the gnome side) a bit more.  You'll note that I've changed the signal to `show`.  Now `transitions-completed` signal did work - but it did throw an exception in the logs re the floating windows' workspace being null - which I posit is due to `transitions-completed` being emitted before the window is rendered (and the workspace property added).

In any case, the `show` signal does not exhibit any errors/exceptions in the log.  The `show` signal is emitted once the actor is fully visible and rendered on the stage.

As for the "correct" workaround - I believe this is it - or at least it's the only known workaround thus far.  Given this, and that it "fixes" (or works around) a pretty severe gnome crashing bug - I say we implement this and if a better one is found in the future we can change it then.

Given that this method is only used for windows on the scratch layer (and simply calls `activate` after it's rendered) - I see there is little chance with causing other issues.

smichel17 commented:
That's enough of an explanation to satisfy me :)
